[clang-cache-build-session] Provide a way to infer the prefix mapping from a CMake build folder

akyrtzi opened the pull request:
A pair of commits:

1. Replaces the `clang-cache-build-session` script with a clang tool executable
2. Introduces `-prefix-map-cmake` flag which instructs `clang-cache-build-session` that it is running within a CMake
    build directory that was enabled for [CMake's File API](https://cmake.org/cmake/help/latest/manual/cmake-file-api.7.html).
    `clang-cache-build-session` will then infer the CMake project's source paths and
    setup `CLANG_CACHE_PREFIX_MAPS` environment variable for `clang-cache` to create prefix mapping
    that results in build and source directory independent compilations.
    
    In order to enable CMake's File API run
    
    ```
    % mkdir -p .cmake/api/v1/query/codemodel-v2
    ```
    
    inside the build directory before invoking the `cmake` configure command; this
    only needs to happen once when the build directory is first created.

akyrtzi committed:
[clang-cache-build-session] Replace the `clang-cache-build-session` script with a clang tool

This enables building more functionality on top of `clang-cache-build-session`.

Note that a minor downside of `clang-cache-build-session` being a tool, instead of a shell script, is that it
requires that shell scripts are annotated with a hashbang (`#!`) at the top, in order to be able to run them.

akyrtzi committed:
[clang-cache-build-session] Provide a way to infer the prefix mapping from a CMake build folder

Introduces `-prefix-map-cmake` flag which instructs `clang-cache-build-session` that it is running within a CMake
build directory that was enabled for [CMake's File API](https://cmake.org/cmake/help/latest/manual/cmake-file-api.7.html).
`clang-cache-build-session` will then infer the CMake project's source paths and
setup `CLANG_CACHE_PREFIX_MAPS` environment variable for `clang-cache` to create prefix mapping
that results in build and source directory independent compilations.

In order to enable CMake's File API run

```
% mkdir -p .cmake/api/v1/query/codemodel-v2
```

inside the build directory before invoking the `cmake` configure command; this
only needs to happen once when the build directory is first created.

akyrtzi commented:
@swift-ci Please test

akyrtzi commented:
@swift-ci Please test

cachemeifyoucan commented:
LGTM in general.

I wonder if `clang` is the correct place for this? Maybe it should be called `llvm-cache-build-session` or just `cache-build-session` and relocated into llvm directory? I think we probably can use the same tool for `swift` or other cctools so it is not just clang (also the tool doesn't depend on anything in clang).

benlangmuir reviewed (commented):
Can we change the way extra source paths work to find a common parent (maybe limited to just the immediate parent)? It's more convenient to have with a single `/^src` mapping for llvm-project than having one for each (sub)project.  If this was transparent to other tools, it wouldn't be such a big deal, but right now you would need to configure lldb yourself.

akyrtzi commented:
> I wonder if `clang` is the correct place for this? Maybe it should be called `llvm-cache-build-session` or just `cache-build-session` and relocated into llvm directory? I think we probably can use the same tool for `swift` or other cctools so it is not just clang (also the tool doesn't depend on anything in clang).

You have a good point, while for `swift` the swift package manager can handle the setup on its own, CMake has support to build swift files so the tool can be useful for `swift` invocations as well.

I'll move to llvm and use `LLVM_CACHE_` prefix for the environment variables.

akyrtzi commented:
> Can we change the way extra source paths work to find a common parent (maybe limited to just the immediate parent)? It's more convenient to have with a single `/^src` mapping for llvm-project than having one for each (sub)project. 

This would be fine for paths within `llvm-project` but I don't think this heuristic should be applied in general. The source paths of a CMake configuration in general could be contained within the same parent or in different parents (when different repos are involved), but the prefix remapping should result in the same remapped paths for both setups.
I considered checking for the presence of `.git` as a way to determine that certain paths are from the same repo but I thought it's better to avoid changing the remapping behavior based on whether the sources are from a cloned repo or not.

> If this was transparent to other tools, it wouldn't be such a big deal, but right now you would need to configure lldb yourself.

I was thinking that as follow-up we could add an option to the tool to print out the remapped paths in a format that you can copy-paste in a `.lldbinit` file, for convenience.

benlangmuir commented:
> The source paths of a CMake configuration in general could be contained within the same parent or in different parents (when different repos are involved), but the prefix remapping should result in the same remapped paths for both setups.

Good point.  It's unfortunate that this produces a worse setup for monorepos.  I suppose if llvm-project itself contained the top-level cmake file it would behave the way I want.

> I was thinking that as follow-up we could add an option to the tool to print out the remapped paths in a format that you can copy-paste in a .lldbinit file, for convenience.

Yeah, that seems like a good idea.

benlangmuir commented on clang/test/CAS/driver-cache-launcher.c:
```diff
@@ -24,6 +24,29 @@
+// RUN: env CLANG_CACHE_CAS_PATH=%t/cas clang-cache-build-session -prefix-map-cmake clang-cache %clang -c %s -o %t.o -### 2>&1 | FileCheck %s -check-prefix=CLANG-CMAKE-PREFIX -DPREFIX=%t
+// RUN: popd
+
+// SESSION-CMAKE-PREFIX: note: setting CLANG_CACHE_PREFIX_MAPS=/llvm/build=/^build;/llvm/llvm-project/llvm=/^src;/llvm/llvm-project/clang=/^src-clang;/llvm/llvm-project/clang-tools-extra=/^src-clang-tools-extra;/llvm/llvm-project/third-party/benchmark=/^src-benchmark;/llvm/llvm-project/other/benchmark=/^src-benchmark-1;/llvm/llvm-project/another/benchmark=/^src-benchmark-2
```
Could break this long check up onto multiple lines with [CHECK-SAME](https://llvm.org/docs/CommandGuide/FileCheck.html#the-check-same-directive).

akyrtzi replied:
`CHECK-SAME` is not checking an exact continuation (it would miss it if an unexpected path showed up in-between).
I tried `CHECK-SAME-NEXT` or `CHECK-NEXT-SAME` and they both worked as `CHECK-SAME`.
Unless I'm missing something, checking the whole line was the way to ensure checking the exact set of paths that should show up.

(though I just realized that I forgot to add `{{$}}` at the end of the line, oops)

benlangmuir replied:
Didn't realize you wanted to check exact match, makes sense.

benlangmuir commented on clang/tools/clang-cache-build-session/ClangCacheBuildSession.cpp:
```diff
@@ -0,0 +1,176 @@
+                         "Accepts a command that triggers a build and, while "
+                         "the command is running,\n"
+                         "all the clang invocations that are part of that "
+                         "build will be sharing the\n"
```
Wording: I suggest "will share" instead of "will be sharing"

akyrtzi replied:
Done 👍

benlangmuir commented on clang/tools/clang-cache-build-session/CMakeFileAPI.cpp:
```diff
@@ -0,0 +1,228 @@
+  if (!Path.startswith(ParentPath))
+    return false;
+  return Path.drop_front(ParentPath.size())
+      .startswith(sys::path::get_separator());
```
We should use `sys::path::is_separator` instead of `get_separator` since there are multiple valid separators on Windows.  And since you already checked ParentPath.size() > Path.size() you know there is at least one character to check.

akyrtzi replied:
Good point, done 👍

akyrtzi committed:
[cache-build-session] Move the `clang-cache-build-session` to llvm as `cache-build-session` tool

This tool is not dependent on clang and could also be useful for `swift` or other cctools during a CMake build.
As part of moving it to llvm, also change the environment variable names to use `LLVM_CACHE_` prefix.

akyrtzi commented:
> I'll move to llvm and use LLVM_CACHE_ prefix for the environment variables.

I pushed an additional commit that does this.

akyrtzi commented:
@swift-ci Please test

akyrtzi commented:
@swift-ci Please test
